src/analysis/bits-bounds-transfer-function.h

```diff
@@ -0,0 +1,367 @@
+#ifndef wasm_analysis_bits_bounds_transfer_function_h
+#define wasm_analysis_bits_bounds_transfer_function_h
+
+#include <unordered_map>
+
+#include "bits-bounds-lattice.h"
+#include "ir/bits.h"
+#include "ir/stack-utils.h"
+#include "stack-lattice.h"
+#include "visitor-transfer-function.h"
+
+namespace wasm::analysis {
+
+class MaxBitsTransferFunction
+  : public VisitorTransferFunc<MaxBitsTransferFunction,
+                               StackLattice<MaxBitsLattice>,
+                               AnalysisDirection::Forward> {
+  MaxBitsLattice& bitsLattice;
+
+public:
+  std::unordered_map<Expression*, Index> exprMaxBounds;
+
+  MaxBitsTransferFunction(MaxBitsLattice& bitsLattice)
+    : bitsLattice(bitsLattice) {}
+
+  void visitConst(Const* curr) {
+    MaxBitsLattice::Element currElement = bitsLattice.getBottom();
+
+    bool addInformation = true;
+    switch (curr->type.getBasic()) {
+      case Type::i32:
+        currElement.setUpperBound(
+          32 - curr->value.countLeadingZeroes().geti32(), curr->value);
+        break;
+      case Type::i64:
+        currElement.setUpperBound(
+          64 - curr->value.countLeadingZeroes().geti64(), curr->value);
+        break;
+      default: {
+        addInformation = false;
+      }
+    }
+
+    if (collectingResults && addInformation) {
+      exprMaxBounds[curr] = currElement.upperBound;
+    }
+
+    currState->push(std::move(currElement));
+  }
+
+  void visitBinary(Binary* curr) {
+    MaxBitsLattice::Element right = currState->pop();
+    MaxBitsLattice::Element left = currState->pop();
+
+    MaxBitsLattice::Element currElement = bitsLattice.getBottom();
+
+    bool addInformation = true;
+    switch (curr->op) {
+      case RotLInt32:
+      case RotRInt32:
+      case SubInt32: {
+        currElement.setUpperBound(32);
+        break;
+      }
+      case AddInt32: {
+        currElement.setUpperBound(
+          std::min(Index(32), std::max(left.upperBound, right.upperBound) + 1));
+        break;
+      }
+      case MulInt32: {
+        currElement.setUpperBound(
+          std::min(Index(32), left.upperBound + right.upperBound));
+        break;
+      }
+      case DivSInt32: {
+        int32_t maxBitsLeft = left.upperBound;
+        int32_t maxBitsRight = right.upperBound;
+        if (maxBitsLeft == 32 || maxBitsRight == 32) {
+          currElement.setUpperBound(32);
+        } else {
+          currElement.setUpperBound(
+            std::max(0, maxBitsLeft - maxBitsRight + 1));
+        }
```

Comment:
Why is this correct? Even if `rightMaxBits` is large, its value at runtime could still be 1, in which case the max bits of the result should be the same as `maxBitsLeft`, right? Same question for unsigned division below.

Reply:
This is what's implementedin `bits.h`. Added a TODO there.